src/org/jgroups/raft/blocks/LockService.java

```diff
@@ -0,0 +1,845 @@
+
+	protected void doUnlock(UUID member, Set<UUID> unlocking) {
+		Set<LockEntry> set = memberLocks.remove(member); if (set == null) return;
+		for (LockEntry lock : set) doUnlock(member, lock, unlocking);
+	}
+
+	protected boolean doUnlock(UUID member, LockEntry lock, Set<UUID> unlocking) {
+		LockStatus prev = HOLDING;
+		UUID holder = null;
+		List<UUID> waiters = null;
+		boolean reset = unlocking != null;
+		if (member.equals(lock.holder)) {
+			do {
+				if (holder != null) {
+					if (waiters == null) waiters = new ArrayList<>(unlocking.size());
+					waiters.add(holder);
+				}
+				holder = lock.unlock();
+			} while (holder != null && unlocking != null && unlocking.contains(holder));
+		} else {
+			prev = lock.waiters.remove(member) ? WAITING : NONE;
+		}
+		if (address.equals(member)) {
+			onCommit(lock.id, prev, NONE, reset);
+		} else if (address.equals(holder)) {
+			onCommit(lock.id, WAITING, HOLDING, reset);
+		}
+		if (log.isTraceEnabled()) {
+			log.trace("[%s] %s unlock %s, prev: %s", address, member, lock.id, prev);
+			if (holder != null)
+				log.trace("[%s] %s lock %s, prev: %s, next: %s", address, holder, lock.id, WAITING, HOLDING);
+		}
+		if (waiters != null) for (UUID waiter : waiters) {
+			unbind(waiter, lock);
+			if (address.equals(waiter)) {
+				onCommit(lock.id, WAITING, NONE, true);
+			}
+			if (log.isTraceEnabled()) {
+				log.trace("[%s] %s unlock %s, prev: %s", address, waiter, lock.id, WAITING);
+			}
+		}
+		return prev != NONE;
+	}
+
+	protected void doReset(List<UUID> members) {
+		Set<UUID> prev = new HashSet<>(memberLocks.keySet());
+		if (log.isTraceEnabled()) {
+			log.trace("[%s] reset %s to %s", address, prev, members);
+		}
+		for (UUID member : members) prev.remove(member);
+		for (UUID member : prev) doUnlock(member, prev);
+	}
+
+	protected LockStatus doQuery(Long lockId, UUID member) {
+		LockEntry lock = locks.get(lockId);
+		LockStatus status;
+		if (lock == null || lock.holder == null) status = NONE;
+		else if (lock.holder.equals(member)) status = HOLDING;
+		else status = lock.waiters.contains(member) ? WAITING : NONE;
+		if (log.isTraceEnabled()) {
+			log.trace("[%s] %s query %s, status: %s", address, member, lockId, status);
+		}
+		return status;
+	}
+
+	protected void bind(UUID member, LockEntry lock) {
+		memberLocks.computeIfAbsent(member, k -> new LinkedHashSet<>()).add(lock);
+	}
+
+	protected void unbind(UUID member, LockEntry lock) {
+		memberLocks.computeIfPresent(member, (k, v) -> {
+			v.remove(lock); return v.isEmpty() ? null : v;
+		});
+	}
+
+	protected void onCommit(long lockId, LockStatus prev, LockStatus curr, boolean reset) {
+		if (prev == curr) return;
+		// In followers, logs and responses of multiple commands can be included in one batch message, since RAFT is
+		// below REDIRECT in protocol stack, the commands applying may occur before the responses completing.
+		if (curr == HOLDING && (prev == WAITING || prev == null && lockStatus.get(lockId) != HOLDING)) {
```

Comment:
This check on the `lockStatus` won't be deterministic. I see the values on that map are added by the client after receiving a response.